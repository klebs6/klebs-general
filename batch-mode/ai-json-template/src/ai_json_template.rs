// ---------------- [ File: ai-json-template/src/ai_json_template.rs ]
crate::ix!();

/// The derived code implements `AiJsonTemplate` for each struct, letting you
/// call `MyStruct::to_template()` to get a JSON “schema” describing how the
/// AI should produce data that matches this layout.
///
pub trait AiJsonTemplate
: Clone 
+ Debug 
+ LoadFromFile 
+ SaveToFile<Error=SaveLoadError> 
+ Serialize 
+ for<'a> Deserialize<'a> 
{
    /// Return a JSON template describing how the AI’s output should be structured.
    /// This might include doc comments or other instructions for each field.
    fn to_template() -> serde_json::Value;
}
