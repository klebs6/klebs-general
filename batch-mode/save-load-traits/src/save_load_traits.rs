// ---------------- [ File: src/save_load_traits.rs ]
crate::ix!();

#[async_trait]
pub trait SaveToFile {

    type Error;

    async fn save_to_file(
        &self,
        filename: impl AsRef<Path> + Send,

    ) -> Result<(), Self::Error>;
}

#[async_trait]
pub trait LoadFromFile: Sized {

    type Error;

    async fn load_from_file(filename: impl AsRef<Path> + Send) 
        -> Result<Self, Self::Error>;

}
