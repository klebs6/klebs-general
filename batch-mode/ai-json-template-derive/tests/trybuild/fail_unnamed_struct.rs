// ---------------- [ File: ai-json-template-derive/tests/trybuild/fail_unnamed_struct.rs ]
// ======================= File: tests/trybuild/fail_unnamed_struct.rs =======================
// A tuple/unnamed struct. The derive macro should fail with an error
// because AiJsonTemplate only supports named structs.

#![allow(dead_code)]

use ai_json_template_derive::*;
use ai_json_template::*;
use serde::{Serialize, Deserialize};
use getset::{Getters, Setters};
use derive_builder::Builder;
use save_load_derive::*;
use save_load_traits::*;

#[derive(SaveLoad,Clone, Debug, Serialize, Deserialize, Getters, Setters, Builder)]
#[derive(AiJsonTemplate)]
#[getset(get = "pub", set = "pub")]
#[builder(setter(into))]
struct FailUnnamedStruct(String, i32);

fn main() {}
