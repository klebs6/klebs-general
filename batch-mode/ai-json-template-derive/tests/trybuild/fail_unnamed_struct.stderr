error: Only support unary struct!
  --> tests/trybuild/fail_unnamed_struct.rs:17:57
   |
17 | #[derive(SaveLoad,Clone, Debug, Serialize, Deserialize, Getters, Setters, Builder)]
   |                                                         ^^^^^^^
   |
   = note: this error originates in the derive macro `Getters` (in Nightly builds, run with -Z macro-backtrace for more info)

error: Only support unary struct!
  --> tests/trybuild/fail_unnamed_struct.rs:17:66
   |
17 | #[derive(SaveLoad,Clone, Debug, Serialize, Deserialize, Getters, Setters, Builder)]
   |                                                                  ^^^^^^^
   |
   = note: this error originates in the derive macro `Setters` (in Nightly builds, run with -Z macro-backtrace for more info)

error: Unsupported shape `unnamed fields`. Expected named fields.
  --> tests/trybuild/fail_unnamed_struct.rs:17:75
   |
17 | #[derive(SaveLoad,Clone, Debug, Serialize, Deserialize, Getters, Setters, Builder)]
   |                                                                           ^^^^^^^
   |
   = note: this error originates in the derive macro `Builder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: AiJsonTemplate derive only supports a struct with named fields or an enum with all unit variants.
  --> tests/trybuild/fail_unnamed_struct.rs:19:1
   |
19 | / #[getset(get = "pub", set = "pub")]
20 | | #[builder(setter(into))]
21 | | struct FailUnnamedStruct(String, i32);
   | |______________________________________^
